crates/cargo-test-macro/src/lib.rs

```diff
@@ -211,18 +214,44 @@ pub fn cargo_test(attr: TokenStream, item: TokenStream) -> TokenStream {
                     continue;
                 }
             }
+            TokenTree::Ident(i) => {
+                // The first time through it will be `fn` the second time is the
+                // name of the test.
+                if test_name.is_none() && num == 1 {
+                    test_name = Some(i.to_string())
+                } else {
+                    num += 1;
+                }
+                ret.extend(Some(TokenTree::Ident(i)));
+                continue;
+            }
             other => {
                 ret.extend(Some(other));
                 continue;
             }
         };
 
-        let mut new_body = to_token_stream(
-            r#"let _test_guard = {
-                let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
-                cargo_test_support::paths::init_root(tmp_dir)
-            };"#,
-        );
+        let name = &test_name
+            .clone()
+            .map(|n| n.split("::").next().unwrap().to_string())
+            .unwrap();
+
+        let mut new_body = if cfg!(windows) {
+            to_token_stream(
+                r#"let _test_guard = {
+                    let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
+                    cargo_test_support::paths::init_root(tmp_dir)
+                };"#,
+            )
+        } else {
+            to_token_stream(&format!(
+                r#"let _test_guard = {{
+                    let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
+                    let test_dir = cargo_test_support::paths::test_dir(std::file!(), "{name}");
+                cargo_test_support::paths::init_root(tmp_dir, test_dir)
+            }};"#
+            ))
+        };
```

Comment:
With an eye towards reducing the amount of duplicated code, I'm wondering if we can try to combine things a little more. So, for example, this could be:


```suggestion
        let mut new_body = to_token_stream(&format!(
            r#"let _test_guard = {{
                let tmp_dir = option_env!("CARGO_TARGET_TMPDIR");
                let test_dir = cargo_test_support::paths::test_dir(std::file!(), "{name}");
                cargo_test_support::paths::init_root(tmp_dir, test_dir)
            }};"#
        ));
```